system/src/Grav/Common/Page/Medium/ImageMedium.php

```diff
@@ -232,31 +218,36 @@ public function srcset($reset = true)
      * @return $this
      */
     public function derivatives($min_width, $max_width, $step = 200) {
-      $width = $min_width;
-
-      // Do not upscale images.
-      if ($max_width > $this->get('width')) {
-        $max_width = $this->get('width');
-      }
-
-      while ($width <= $max_width) {
-        $ratio = $width / $this->get('width');
-        $derivative = MediumFactory::scaledFromMedium($this, 1, $ratio);
-        if (is_array($derivative)) {
-          $this->addDerivative($derivative['file']);
+        $width = $min_width;
+
+        // Do not upscale images.
+        if ($max_width > $this->get('width')) {
+            $max_width = $this->get('width');
         }
-        $width += $step;
-      }
-      return $this;
-    }
 
-    /**
-     * Add a derivative
-     *
-     * @param  ImageMedium $image
-     */
-    public function addDerivative(ImageMedium $image) {
-      $this->derivatives[$image->url()] = $image->get('width');
+        // Clear any alternatives that have already been implied by the image's
+        // filename (eg. ones that had "@2x" or similar appended to their
+        // filenames)
+        $this->reset();
+
+        while ($width <= $max_width) {
+            $ratio = $width / $this->get('width');
+            $width += $step;
+
+            if ($ratio == 1) {
+                continue;
+            }
+
+            $derivative = MediumFactory::fromFile($this->get('filepath'));
+            $derivative_width = $derivative->get('width') * $ratio;
```

Comment:
Alright, getting through this was a real ordeal, but it looks like I figured out what the issue was.

At first I thought it was PHP that was returning errant results when listing the contents of a directory, but it turns out that it was Grav's page object caching mechanism that didn't pick up on updates to just the media files of a page object. I described this issue in #1080 as well. 

`Media::__construct` (which initializes a page's media), is allowed to continue even if `MediumFactory::fromFile` returns null. This patch doesn't take that risk into account, and IMO, it shouldn't either. If the original `ImageMedium` was initiated, it should reliably point to an actual file.

Would be awesome if we could work through this final issue and get this PR merged, @rhukster!